### PR TITLE
udp_session_filters: dynamic forward proxy session filter

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description: Adds a UDP session filter that's able to perform DNS resolution for dynamic host based on filter state values. Using a custom user's session filter placed before the dynamic forward proxy filter, it's able to have the same functionality for UDP flows similar to TCP and HTTP. This was originated by #29277, although not really needed to perform tunneling.
Risk Level: low, new extension
Testing: unit tests, integration tests
Docs Changes: change log, API docs, configuration docs
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

**Comment by ohadvano** [comment]
@mattklein123  I think Lizan is out, would you mind doing API and code pass as well, since your code owner?

**Comment by ohadvano** [comment]
ping @lizan, @mattklein123 
cc @alyssawilk

**Comment by mattklein123** [comment]
Sorry no I don't have time to review this.

**Comment by ohadvano** [comment]
Thanks for the update Matt. @alyssawilk can you please help assign relevant reviewers to help this PR progress? I see you also had some DFP involvment, could you maybe take this one?

**Comment by ohadvano** [comment]
@ravenblackx for some reason I thought Alyssa was the on-call this week (therefore previous comment). Can you please assign this PR to the relevant reviewers?

**Comment by ravenblackx** [comment]
@adisuissa can you api-shepherd this?
@rshriram as [code-owner of dynamic_forward_proxy](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/CODEOWNERS?L55), any chance of a review?

**Comment by ohadvano** [comment]
Ping @adisuissa

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc:181]: Super nit, could you name the tests a bit more descriptively of what's being tested (e.g. `PassesThroughImmediatelyWhenDnsAlreadyInCache` or something like that, in this case; request applies to most test cases)

It makes it easier to see if the tests are checking what they're supposed to, if the name describes what the test is supposed to check.

**Review by ravenblackx — CHANGES_REQUESTED** [review]

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc:245]: Might be worth validating the stat-write in this test, I don't think it's validated anywhere else.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:177]: This is derived from the ``LoadBalancerContext`` interface, and its method: ``const StreamInfo::StreamInfo* requestStreamInfo() const``. If you look at the ``UdpLoadBalancerContext``, it overrides this method so it's required to provide a pointer here

File: api/envoy/extensions/filters/udp/udp_proxy/session/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/udp/udp_proxy/session/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto:24]: Please add a link to the filter that sets these filter state keys.

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ohadvano) [reply test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc:245]: I think it may be hard testing it in a unit test, but it's validated in the integration test

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ohadvano) [reply test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc]: Done, thanks

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ohadvano) [reply test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc]: done

**Review by adisuissa — COMMENTED** [review]
Thanks!
High-level API questions/comments:
1) Is this a udp-only filter? Wouldn't there be a need for a tcp-proxy one as well?
If the config is a bit different, I suggest writing the shared parts in a "common" file, and the udp part can be kept here
2) Will something similar to http's DFP [implementation_specifier](https://github.com/envoyproxy/envoy/blob/4d46da0bba54dfb849d8bf68b600e53d87310a1a/api/envoy/extensions/filters/http/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto#L28) will be used? If so, should probably add the oneof now.
3) Seems that this can benefit from better docs with an example (see [example](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/dynamic_forward_proxy_filter))
4) OOC why is the directory structure under "filters/udp" rather than "filters/network/udp"?

File: source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc
Comment (jmarantz) [reply source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc:27]: We definitely shouldn't create counters this way, but creating a scope name this way in the config path is not strictly problematic -- we are not too worried about lock contention as config-creation happens in the main thread on xds update. There is a more robust mechanism in https://github.com/envoyproxy/envoy/blob/4d46da0bba54dfb849d8bf68b600e53d87310a1a/source/common/stats/utility.h#L95C16-L95C34 but it wouldn't necessarily be easier to use in this case, unless you have StatNames saved in a factory or something for all 3 segments of "udp.session.dynamic_forward_proxy".

Several questions though:
 * "udp.session.dynamic_forward_proxy" is a lot of prefix. Do you really need all that?
 * what is the name of context.scope() ? That will be prefixed to whatever you specify.
 * why use fmt::format? absl::StrCat should be faster for this.

File: api/envoy/extensions/filters/udp/udp_proxy/session/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto
Comment (ohadvano) [reply api/envoy/extensions/filters/udp/udp_proxy/session/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto:24]: 1. Yes this one is a UDP **session** filter only
2. Relevant to previous question too - so Envoy has ``sni_dynamic_forward_proxy`` for network filters, and ``dynamic_forward_proxy`` for HTTP filters. This one adds a similar functionality but for UDP session filters.
3. I'm not sure this doc would be relevant to the UDP variation of the filter, but I indeed added a reference to the [dynamic forward proxy architecture overview](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_proxy#arch-overview-http-dynamic-forward-proxy) in the filter docs.
4. Don't have a solid answer here, but possible to distinguish between the kinds of filters and where they can be applied.

> Please add a link to the filter that sets these filter state keys.

There is no filter that sets these filter state keys. The idea is that whoever uses this filter, needs to provide a custom filter that sets the values as they required for their use case - this is specified in the docs I added in this PR. Alternatively, a user could theoretically use Kuat's new filter once it's available #29844.

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ravenblackx) [inline test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc:195]: This might be a useful name if you already know what's supposed to happen, but...

How about `RequestWithCacheMissShouldStopIterationBufferPopulateCacheAndFlushWhenReady`

I realize it's a ridiculously long name, but I think all these things are being tested here, and it would be nice to know that that's the intent of the test case.

**Review by ravenblackx — APPROVED** [review]

File: test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc
Comment (ravenblackx) [reply test/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter_test.cc:245]: Ah, you're right, I missed that (I thought I'd seen it, but when I searched for 'stat' in the test files it wasn't there - guess I needed to look manually). Thanks!

File: source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/session_filters/dynamic_forward_proxy/proxy_filter.cc]: >"udp.session.dynamic_forward_proxy" is a lot of prefix. Do you really need all that?

In my perspective is makes sense ``udp.session`` is the type of the filter and ``dynamic_forward_proxy`` is the name of the filter. Do you have another suggestion?

> what is the name of context.scope() ? That will be prefixed to whatever you specify.

I believe there's no other prefix as it's provided from the factory. The integration tests verify this.

> why use fmt::format? absl::StrCat should be faster for this.

Done, thanks

**Review by ravenblackx — DISMISSED** [review]

**Comment by ohadvano** [comment]
@adisuissa I didn't read your 2. comment correctly initially, adjusted the API now, thanks. PTAL

**Comment by ohadvano** [comment]
Ping @adisuissa

**Review by ravenblackx — APPROVED** [review]

**Comment by jmarantz** [comment]
I'm assuming Raven's earlier approval still stands as the intervening commit was minor. However this needs a senior maintainer approval.

@mattklein123 I'm not sure if you have time but you might have the most context on dynamic forward proxy filters. If you are not able to do it ping me and I'll re-assign to a different senior maintainer.

**Review by ravenblackx — APPROVED** [review]

**Comment by mattklein123** [comment]
Yup I can take a look through in the next couple of days.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
